### DISCUSSION TITLE
Load Test Leads to 400 error

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#5321] [Q&A]
I have simulated load test by but lot of request leads to 400 error in logs i found below as error
<img width="621" alt="image" src="https://github.com/user-attachments/assets/b69d53ff-69af-42d3-bef2-62be14acb773">

Test Configuration given below
<img width="539" alt="image" src="https://github.com/user-attachments/assets/e39b269c-26f6-48f2-8e6a-d6d68044be77">

Result of load test given below 
<img width="1193" alt="image" src="https://github.com/user-attachments/assets/4f5cf59b-d795-4d7c-b19d-0206e6729719">

I am wondering nearly half of request leads to error. some after hitting pocket base and some timeout which is more than 50 sec.
Server spec is 16 vcpu & 16GB RAM

@ganigeorgiev any suggestion on optimisation to get better throughput?

**Comment by ganigeorgiev** [comment]
This seems to be some custom query. Without details what the load script actually does or what is your custom hooks code there is nothing much I can help here.

**Reply by iamdineshkumar** [reply]
Load is much straightforward simple authentication
![image](https://github.com/user-attachments/assets/f1515226-d457-4877-8947-18315c0ab16a)

Note: its freshly deployed DB

**Reply by ganigeorgiev** [reply]
Sorry, without clear reproducible steps I don't think I can help you.

If you can't reproduce the issue locally then most likely there is something with your deployment setup.

Additionally note that the `auth-with-password` calls are expected to be slow due to the bcrypt hash generation.

For benchmarks you could check https://github.com/pocketbase/benchmarks.

If you think that there is a PocketBase issue, feel free to provide more details or some steps to reproduce and I'll try to investigate it but I'm not going to invest time on a wild goose chase.

**Reply by iamdineshkumar** [reply]
I have used https://loader.io/ service free plan

1. Added host in loader.io and got verification token details
2. to verify host added token file in public directory of pocketbase
3. Deployed PB In server and created single user from pb admin login.
4. Then verified host and created test with Test type as "Maintain client load" clients range to 10000 and on advance just increased time out to 50 sec to check higher response time as well.
5. Configured client request as in screenshot shared and ran test

Above is the step i did. So haven't customised anything or not integrated any third party code. No migration just a vanilla PB v0.22.18

**Reply by ganigeorgiev** [reply]
Fine, I've decided to waste some time and run the test for the live demo - pocketbase.io.

The live demo is using the smallest Intel Hetzner CX22 instance (2 shared vCPU, 4GB RAM).

The test was against `GET https://pocketbase.io/api/collections/messagesReport/records`.
I've decided on the `messagesReport` view collection because the query is the most "complex" one in the demo since it contains `COUNT` and `GROUP BY` aggregation.

Below are the results:

![test_config](https://i.imgur.com/1rnvS9b.png)
![times_screenshot](https://i.imgur.com/jF4eT2B.png)
![details_screenshot](https://i.imgur.com/IEGv064.png)

_(**edit**: I've also just noticed that the request the test is performing was over http and the above result actually includes also extra http->https redirect)_

Is this OK?

I don't know, it depends what are your expectations and desired outcome.

For me it seems reasonably OK considering that this is stock PocketBase with the default 120 max open db connections. I didn't even bother increasing the process file descriptors limit of ~4000 (which is most likely the reason for the few "failed to open db" 400 errors).

Furthermore the test performance could be affected by many other reasons:

- client-server locations
- provider's network level protections like DDoS, requests throttling, rate limits, etc. (especially if the requests originate from the same IP as I assume is the case with the free load testing tool)
- provider's hardware protections like CPU and disk throttling (especially if you use shared CPU and network mounted volumes) to prevent mining and other crypto abuse (this is also why the bcrypt password verification could be delayed further)
- the test execution environment (ex. internet speed)

So at the end it really depends what actually is the end goal here and what you are trying to verify with the test.

In general, if you are expecting large number of concurrent requests and you are getting "file open errors" or similar, you'll have to increase the file descriptors limit either as part of the systemd service file or via `ulimit` (see https://pocketbase.io/docs/going-to-production/#increase-the-open-file-descriptors-limit).
Increasing the default number of [max allowed db connections](https://pkg.go.dev/github.com/pocketbase/pocketbase#Config) to something bigger could also eventually help with the test performance if that is your goal, but in real live scenario it could also have negative impact since it will increase the memory usage and eventually causing OOM errors (with the refactoring the default limit is increased from 120 to 250 but I haven't done any tests yet on the upcoming version so I may have to adjust it further to something "safer").

Can PocketBase performance be improved - sure, but unless a critical issue or bottleneck arise for now I tend to prioritize the general DX and correctness over performance. Performance improvements and detailed memory allocation profiling are left for last (before or after stable v1).

Again, I'm willing to investigate if you think that there is a PocketBase issue but without anything concrete I don't know how to help further.

**Comment by iamdineshkumar** [comment]
I have done same for list API with 10 records and with guest access as rule without auth token results are given below. This time there is no much errors. But those errors where 500 for which logs not available.

<img width="1155" alt="image" src="https://github.com/user-attachments/assets/94997a94-2eb7-443f-af82-774e846e2fee">
